torch/_dynamo/output_graph.py

```diff
@@ -511,10 +515,14 @@ def register_attr_or_module(
         assert not isinstance(source, ParamBufferSource)
 
         if isinstance(target, torch.Tensor):
-            if len(self.tracers) > 1:
-                unimplemented(
-                    "accessing attribute of nn.Module inside HigherOrderOperator"
-                )
+            tracer = self.current_tracer
+            if not self.is_root_tracer():
+                # For higher order ops, we don't want to insert the get_attr in
+                # innermost graph. Instead, we want to raise the params/buffers
+                # as inputs to the higher-order graph, and register them as
+                # get_attrs in the root tracer.
+                tracer = self.root_tracer
```

Comment:
Hmm I am not sure about this - we already do lifting through `lift_tracked_freevar_to_input` - why doesn't this go through that?

Reply:
The main difference is that we don't want to raise the parameters as inputs of the outer graph. This way, we don't change the calling convention for downstream users.

Reply:
Note that Dynamo will still call `lift_tracked_freevar_to_input` for the inner graph. The only difference is what happens at the root.  If not a parameter, this is what happens today for higher order ops

https://github.com/pytorch/pytorch/blob/2fa1b563da4b14c29e1a8572cefc79280caf7764/torch/_dynamo/variables/builder.py#L814-L824


This is resulting graph

~~~
 ===== __compiled_fn_0 =====
 <eval_with_key>.1 class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_

        # File: /scratch/anijain/work/pytorch/torch/nn/modules/linear.py:114, code: return F.linear(input, self.weight, self.bias)
        g__mod___linear_weight = self.G__mod___linear_weight
        g__mod___linear_bias = self.G__mod___linear_bias

        # File: /scratch/anijain/work/pytorch/examples/higher.py:39, code: return torch.utils.checkpoint.checkpoint(mod, x)
        wrap_body_0 = self.wrap_body_0
        wrap_activation_checkpoint = torch._higher_order_ops.wrap.wrap_activation_checkpoint(wrap_body_0, l_x_, g__mod___linear_weight, g__mod___linear_bias);  wrap_body_0 = l_x_ = g__mod___linear_weight = g__mod___linear_bias = None
        return (wrap_activation_checkpoint,)

    class GraphModule(torch.nn.Module):
        def forward(self, l_x_, g__mod___linear_weight, g__mod___linear_bias):
            # File: /scratch/anijain/work/pytorch/torch/nn/modules/linear.py:114, code: return F.linear(input, self.weight, self.bias)
            linear = torch._C._nn.linear(l_x_, g__mod___linear_weight, g__mod___linear_bias);  l_x_ = g__mod___linear_weight = g__mod___linear_bias = None

            # File: /scratch/anijain/work/pytorch/examples/higher.py:34, code: return torch.sigmoid(self.linear(x))
            sigmoid = torch.sigmoid(linear);  linear = None
            return sigmoid
~~~